Keeping track of fetched packages and commits to avoid duplicates and make fetching remotes faster

mihem opened the pull request:
@b-rodrigues My inital plan was to improve my existing code to remove duplicates so that your post-processing function does even has to do anything
However, I then realized that this was 1) pretty complex 2) quite inefficient because I call `get_remote` to get the remote dependencies for each remote dependencies. This mean for our example with `mlr3extralearners` GithubAPIs call for each of the 6 remote dependencies.
That's why i decided to remove my previous added code.

So with the main branch this takes ~30 on my machine
```
library(rix)

my_token <- gitcreds::gitcreds_get()$password
Sys.setenv(GITHUB_PAT = my_token)

system.time(
  rix(
    date = "2023-12-30",
    r_pkgs = c(
      "tidyverse", "mlr3", "qs", "matrixStats", "prospectr", "Cubist",
      "checkmate", "mlr3misc", "paradox"
    ),
    system_pkgs = NULL,
    git_pkgs = list(
      package_name = "mlr3extralearners",
      repo_url = "https://github.com/mlr-org/mlr3extralearners/",
      commit = "6e2af9ef9ecd420d2be44e9aa2488772bb9f7080"
    ),
    ide = "none",
    project_path = ".",
    overwrite = TRUE,
    print = TRUE
  )
)
```

```
  user  system elapsed 
  2.182   0.913  30.446 
```

With the new branch the same code takes 23s
```
   user  system elapsed 
  1.623   0.777  23.639 
```

There are also no duplicates.

This also works fine with this code (my inital try to remove duplicates): https://github.com/ropensci/rix/pull/385

However, this leads to a double entry of seurat-data: https://github.com/ropensci/rix/pull/388

So to be explicit:


```
rix(
  r_ver = "4.4.1",
  r_pkgs = NULL,
  system_pkgs = NULL,
  git_pkgs = list(
    package_name = "seurat-wrappers",
    repo_url = "https://github.com/satijalab/seurat-wrappers",
    commit = "5b63c23b211002611b0ccc33da009237d0dbafb6"
  ),
  ide = "none",
  project_path = ".",
  overwrite = TRUE,
  print = TRUE
)
```

has `seurat-data` twice.

```
let
 pkgs = import (fetchTarball "https://github.com/rstats-on-nix/nixpkgs/archive/2024-10-01.tar.gz") {};
  
    CIPR-Package = (pkgs.rPackages.buildRPackage {
      name = "CIPR-Package";
      src = pkgs.fetchgit {
        url = "https://github.com/atakanekiz/CIPR-Package";
        rev = "8729f6bda9f9a178f77844cab9467904fb9ecdae";
        sha256 = "sha256-oeVNSIig3+1dG6sjAxdvRVzDOfIcPYSp8psaQepDoG8=";
      };
      propagatedBuildInputs = builtins.attrValues {
        inherit (pkgs.rPackages) 
          ggpubr
          gtools
          tibble
          dplyr
          rlang
          magrittr;
      };
    });


    conos = (pkgs.rPackages.buildRPackage {
      name = "conos";
      src = pkgs.fetchgit {
        url = "https://github.com/hms-dbmi/conos";
        rev = "c3c1a941548e3bca1f4731f18b2abe16d126cdf3";
        sha256 = "sha256-7H/O+zDHBlQzU+hWYESin4yHitLqIVOrQOcvrTiJz/c=";
      };
      propagatedBuildInputs = builtins.attrValues {
        inherit (pkgs.rPackages) 
          Matrix
          igraph
          abind
          cowplot
          ComplexHeatmap
          dendextend
          dplyr
          ggplot2
          ggrepel
          gridExtra
          irlba
          leidenAlg
          magrittr
          Matrix_utils
          N2R
          R6
          reshape2
          rlang
          Rtsne
          sccore
          Rcpp
          RcppArmadillo
          RcppEigen
          RcppProgress;
      };
    });


    harmony = (pkgs.rPackages.buildRPackage {
      name = "harmony";
      src = pkgs.fetchgit {
        url = "https://github.com/immunogenomics/harmony";
        rev = "ea48a6c3c67f6c4c33841b3f545412d38ddcb363";
        sha256 = "sha256-BhGY+aG5vmNNw1LRei9S010cRbyoTcwwNnPRlAElnxw=";
      };
      propagatedBuildInputs = builtins.attrValues {
        inherit (pkgs.rPackages) 
          R
          Rcpp
          dplyr
          cowplot
          tidyr
          ggplot2
          irlba
          Matrix
          tibble
          rlang
          SingleCellExperiment
          RcppArmadillo
          RcppProgress;
      };
    });


    liger = (pkgs.rPackages.buildRPackage {
      name = "liger";
      src = pkgs.fetchgit {
        url = "https://github.com/welch-lab/liger";
        rev = "8ee61cafd28cce6aaf62fd68acdc4dabb857b59f";
        sha256 = "sha256-4O9SLFZ0nad0j8sR2dKtEBcpbLWkeFAGYMtwhZuzpmU=";
      };
      propagatedBuildInputs = builtins.attrValues {
        inherit (pkgs.rPackages) 
          cowplot
          Matrix
          patchwork
          Rcpp
          plyr
          FNN
          dplyr
          ggrepel
          irlba
          ica
          Rtsne
          ggplot2
          riverplot
          foreach
          doParallel
          mclust
          psych
          RANN
          uwot
          rlang
          hdf5r
          scattermore
          RcppArmadillo
          RcppEigen
          RcppProgress;
      };
    });


    monocle3 = (pkgs.rPackages.buildRPackage {
      name = "monocle3";
      src = pkgs.fetchgit {
        url = "https://github.com/cole-trapnell-lab/monocle3";
        rev = "87f6e88760566065179ae5039d524da7d032cc15";
        sha256 = "sha256-QzGXPDXIMOE5s+cqJQhSMgFx1QXdVoWNXeXfcu/Eblc=";
      };
      propagatedBuildInputs = builtins.attrValues {
        inherit (pkgs.rPackages) 
          Biobase
          SingleCellExperiment
          assertthat
          batchelor
          BiocGenerics
          BiocParallel
          DelayedArray
          DelayedMatrixStats
          digest
          dplyr
          future
          ggplot2
          ggrastr
          ggrepel
          grr
          HDF5Array
          igraph
          irlba
          leidenbase
          limma
          lme4
          lmtest
          MASS
          Matrix
          Matrix_utils
          openssl
          pbapply
          pbmcapply
          pheatmap
          plotly
          plyr
          proxy
          pscl
          purrr
          RANN
          RColorBrewer
          Rcpp
          reshape2
          rsample
          RhpcBLASctl
          RcppAnnoy
          RcppHNSW
          Rtsne
          S4Vectors
          shiny
          slam
          spdep
          speedglm
          stringr
          SummarizedExperiment
          uwot
          terra
          tibble
          tictoc
          tidyr
          viridis;
      };
    });


    Nebulosa = (pkgs.rPackages.buildRPackage {
      name = "Nebulosa";
      src = pkgs.fetchgit {
        url = "https://github.com/powellgenomicslab/Nebulosa";
        rev = "b834a75c5dc28c88d44a640cad5e1cb88b94e138";
        sha256 = "sha256-3XyK6dcMhi15bpmE0FT/L5ssI4xTK6l8WZdVaQwhrwM=";
      };
      propagatedBuildInputs = builtins.attrValues {
        inherit (pkgs.rPackages) 
          ggplot2
          patchwork
          SingleCellExperiment
          SummarizedExperiment
          SeuratObject
          ks
          Matrix;
      };
    });


    presto = (pkgs.rPackages.buildRPackage {
      name = "presto";
      src = pkgs.fetchgit {
        url = "https://github.com/immunogenomics/presto";
        rev = "052085db9c88aa70a28d11cc58ebc807999bf0ad";
        sha256 = "sha256-aEJQThmKdFprMSZRxle+gbvd0mS+077XpuKSTJ6qbrA=";
      };
      propagatedBuildInputs = builtins.attrValues {
        inherit (pkgs.rPackages) 
          Rcpp
          data_table
          dplyr
          tidyr
          reshape2
          Matrix
          rlang
          DESeq2
          RcppArmadillo;
      };
    });


    schex = (pkgs.rPackages.buildRPackage {
      name = "schex";
      src = pkgs.fetchgit {
        url = "https://github.com/SaskiaFreytag/schex";
        rev = "031320d";
        sha256 = "sha256-fd46ZxXEEk2XsEhQExrvpx+l5GPwGSrzBLTrJQIUVvg=";
      };
      propagatedBuildInputs = builtins.attrValues {
        inherit (pkgs.rPackages) 
          SingleCellExperiment
          Seurat
          ggplot2
          hexbin
          scater
          cluster
          dplyr;
      };
    });


    seurat-data = (pkgs.rPackages.buildRPackage {
      name = "seurat-data";
      src = pkgs.fetchgit {
        url = "https://github.com/satijalab/seurat-data";
        rev = "d6a8ce61ccb21a3b204f194d07009772c822791d";
        sha256 = "sha256-DMVbzWJdgNj1sPRLwFToa6O4I0bTc/ECgtg3Cg/o+jc=";
      };
      propagatedBuildInputs = builtins.attrValues {
        inherit (pkgs.rPackages) 
          cli
          crayon
          rappdirs
          Matrix
          Seurat
          SeuratObject;
      };
    });


    seurat-data = (pkgs.rPackages.buildRPackage {
      name = "seurat-data";
      src = pkgs.fetchgit {
        url = "https://github.com/satijalab/seurat-data";
        rev = "b198ed523b1fd773c09356d30d9dfcebd62f22cf";
        sha256 = "sha256-9fGp5dRzfPO7/vh8yuVq9YbDQLRkYEBL+Z6u2J2BFnw=";
      };
      propagatedBuildInputs = builtins.attrValues {
        inherit (pkgs.rPackages) 
          cli
          crayon
          rappdirs
          Matrix
          Seurat
          SeuratObject;
      };
    });

    seurat-disk = (pkgs.rPackages.buildRPackage {
      name = "seurat-disk";
      src = pkgs.fetchgit {
        url = "https://github.com/mojaveazure/seurat-disk";
        rev = "9b89970eac2a3bd770e744f63c7763419486b14c";
        sha256 = "sha256-NqhABs5KsqWiMckG/l3fXzra+DnE7K2HpdtVtyxxr2I=";
      };
      propagatedBuildInputs = builtins.attrValues {
        inherit (pkgs.rPackages) 
          cli
          crayon
          hdf5r
          Matrix
          R6
          rlang
          Seurat
          SeuratObject
          stringi
          withr;
      } ++ [ seurat-data ];
    });


    velocyto.R = (pkgs.rPackages.buildRPackage {
      name = "velocyto.R";
      src = pkgs.fetchgit {
        url = "https://github.com/velocyto-team/velocyto.R";
        rev = "83e6ed92c2d9c9640122dcebf8ebbb5788165a21";
        sha256 = "sha256-5iVHDHsN9xvD54nbva6SC0vVR4R3xXBi2ctM88eE7+I=";
      };
      propagatedBuildInputs = builtins.attrValues {
        inherit (pkgs.rPackages) 
          Matrix
          Rcpp
          MASS
          mgcv
          pcaMethods
          abind
          cluster
          igraph
          hdf5r
          RcppArmadillo;
      };
    });

    seurat-wrappers = (pkgs.rPackages.buildRPackage {
      name = "seurat-wrappers";
      src = pkgs.fetchgit {
        url = "https://github.com/satijalab/seurat-wrappers";
        rev = "5b63c23b211002611b0ccc33da009237d0dbafb6";
        sha256 = "sha256-M7KJrP70z81U9j7wva/HWipgmEwXKv4Us8IuP1SRPUU=";
      };
      propagatedBuildInputs = builtins.attrValues {
        inherit (pkgs.rPackages) 
          BiocManager
          cowplot
          ggplot2
          igraph
          Matrix
          remotes
          rsvd
          Seurat
          rlang
          R_utils
          data_table
          reticulate;
      } ++ [ liger conos harmony presto seurat-data velocyto.R schex monocle3 seurat-disk Nebulosa CIPR-Package ];
    });
    
  system_packages = builtins.attrValues {
    inherit (pkgs) 
      glibcLocales
      nix
      R;
  };
  
in

pkgs.mkShell {
  LOCALE_ARCHIVE = if pkgs.system == "x86_64-linux" then "${pkgs.glibcLocales}/lib/locale/locale-archive" else "";
  LANG = "en_US.UTF-8";
   LC_ALL = "en_US.UTF-8";
   LC_TIME = "en_US.UTF-8";
   LC_MONETARY = "en_US.UTF-8";
   LC_PAPER = "en_US.UTF-8";
   LC_MEASUREMENT = "en_US.UTF-8";

  buildInputs = [ seurat-wrappers   system_packages   ];
  
}

```

Do you know why? Only one thing I realized is that the `ref` of the seurat-data is different. So Maybe this should only check for the 

And second thing is that if packages are removed there are lots of blank lines (now the mlr3extralearners example again). 
[default.nix.log](https://github.com/user-attachments/files/18811666/default.nix.log)

mihem commented:
The fact that the dups tests fails is only because of differences in blank lines. It would be nice if we could create a clean version without additional blank lines thatn we don't have to worry about that.

b-rodrigues commented:
if there are cases where duplicates remain, I prefer to keep the coding as is to be honest

mihem commented:
haha i hear you. This issue is quite challenging and can be annoying. 
In general I think it would be better if we can remove code, make it more efficient,and not use two half working approaches.
But sure, it's working now. I will try to see if I can fix those problems. If I don't manage it, we just keep it as it is, agree.

b-rodrigues commented:
consecutive empty lines are now removed with https://github.com/ropensci/rix/commit/4b31d109db287a7fa16a921292f83ef4a86b2a5d

mihem commented:
nice.
Wow the function `remove_emptry_lines` is nice, so short, fast and powerful, but not easy to grasp at first sight (at least for my mind).

I am experimenting with a solution to keep track of the already fetched packages using environments (Claude Sonnet 3.5 suggestion) btw, but not there yet

b-rodrigues commented:
>Wow the function remove_emptry_lines is nice, so short, fast and powerful, but not easy to grasp at first sight (at least for my mind).

we can thank chatgpt for it 😂

mihem commented:
Yeah, we can only thank Github Copilot (Claude Sonnet 3.5) for the idea with environemnts.
I think that's exactly what I had tried to achieve for a few weeks :) 
I went throught the example of mle3extralearners with `skip_post_processing = TRUE` and there were not duplicates and magically I could even build the nix-shell (so no version mismatch as before), although this was probably more luck?

Also the seurat-wrappers example worked fine (no duplicates).

So if you have time, have a look, I haven't used this environment approach, it's a little awkward, but also quite cool because don't know how to carry the already fetched packages through the functions.

But don't think this is ready to merge, needs some more testing.

mihem commented:
hm, but it changed the commit hash, and now it's not the closest anymore, have to further investigate

@b-rodrigues btw the ubuntu-checks are still not showing if they fail. I know have other commit hash, so there are 2 tests failing in devtools_test (ubuntu-latest), but it says it ran succesfully. I think it would be helpful if you could fix that.

mihem commented:
just furthing thinking out loud:

So mlre3xtralearners with the given commit hash `6e2af9ef9ecd420d2be44e9aa2488772bb9f7080` has the commit date 2023-10-11. And then rix fetches the commit hash `a901255c26614a0ece317dc849621420f9393d42` for `set6` (commit date 2023-09-01.  `param6` closet commit to 2023-10-11 is `0fa35771276fc05efe007a71bda466ced1e4c5eb` which is very old 2022-08-28. The problem is that `set6` is also a dependency of `param6` so it's fetched again (note: this is even though we now have an environment that keeps track of the already fetched packages because `param6` could have further dependencies that have not been fetched yet so it makes sense to in general check for that), and now the closest commit to the date 2022-08-28 is `e65ffeea48d30d687482f6706d0cb43b16ba3919` of 2022-08-27.

Not sure there is a right version here. We are probably again at the point where the user has to decide, right @b-rodrigues. But this is the reason why the tests fail and why on the other hand building the shell now works.

I think it would more sense to skip fetching the package again (so i think we need another environment that keeps track of that) because this is faster and then we should end up with all dates being close to the date of the given commit, although this might also require manual intervention. @b-rodrigues agree?

b-rodrigues commented:
I haven't had the chance to look at this if depth but I generally don't like when functions mess with the global environment, isn't it possible to keep the track in a variable within the scope of the function ?

mihem commented:
Yeah I know that this is not super clean, but we can remove that variables again from the global environment.

I don't think it's that easy. That's because of this recursive call offetchgits within fetchgit.

```
  } else { # if there are remote dependencies, start over

    remote_packages_expressions <- fetchgits(remotes)
```

So then we are in fetchgit with e.g. `param6` as a dependency of `mlr3extralearners` and all information about prior fetched packages is gone.

b-rodrigues commented:
wouldn't adding a variable `seen_pkgs ` to ftchgit with a default set to NULL do the trick? I don't want rix() to be writing to the global env

mihem commented:
I don't think so because when fetchgits is called it's a new environment then and `seen_pkgs` is gone.
Not an expert concerning environments, are there other environments we could use?
It's exactly the solution I was looking for and when I am debugging this rix calls it's pretty cool because it now really removes all duplicate packages and all tests have worked so far.

b-rodrigues commented:
why not cache the list of seen packages in a /tmp/?

mihem commented:
git commits were quite a mess, not it's all rebased

okay i try to find a solution to avoid global environments. I tried several solutions, but none of them worked so far. I used s3 objects, however also tricky to really keep values across functions, that would at last mean that i have to carry this object through all functions (from fetchgits to resolve_package_commit are quite some). I also tried your solution, which seemed to work fine. But i then realized that for each loop it creates a new temporary file ..  right now experimenting with environments , but not global environments, because so far i have found this easiest to handle.
Quite a lot of work, but I also learn a lot :)

mihem commented:
Hehe, okay in the end I used your solution and saved it to a temporary file (avoids creating a new file by using fixed file name, not that complicated).
At least in my interactive tests it worked fine and no duplicates with the two tests rix calls (seurat-wrappers and mlr3extralearners).
But need to do further testing.

THe dedup tests fails because of some format stuff and because deup actually has the "wrong" commit (meaning of a date   2022-08-27 as explinaed above)

mihem commented:
I am getting closer. It's pretty annoying with the snapshots because each time we change some code they change a little and it's sometimes quite difficult to get the exact same.

Please keep in mind to change the link in  `test-dedup.yaml`
           wget -O dups-entries_default.nix https://raw.githubusercontent.com/mihem/rix/refs/heads/remove_duplicate_third/tests/testthat/_snaps/rix/dups-entries_default.nix

mihem commented:
Nice now all tests pass  :100: 
So I think this is ready for your revision.

b-rodrigues commented:
any way to test if the cache actually works? because assuming it somehow didn't, the function would still go through right, it'd just take longer?

mihem commented:
Ah okay you are right, I haven't thought that much about failing yet because unlike using GithubAPI, saveRDS and readRDS should always work.
What would you suggest what should be tested? What would you expect could fail?

b-rodrigues commented:
I'm not sure, but I think that we would need to handle failure

mihem commented:
Okay and do you have anything in mind which failure exactly?

b-rodrigues commented:
> Sorry, I just have to run this locally and then ..?

you should be prompted to confirm my email address

mihem commented on R/fetchers.R:
```diff
@@ -377,12 +370,23 @@ fetchlocals <- function(local_r_pkgs) {
 #' @noRd
 fetchgits <- function(git_pkgs) {
+  cache_file <- get_cache_file()
+  cache <- readRDS(cache_file)
```
hm I used `on.exist(unlink)` in a first attempt. However, this did not work because it mean the cache file was deleted too early. I can check if I can include this somewhere, where it does not get deleted before all fetching is done, but that's why i put it into the rix main function.

mihem replied:
this did not work when i put it in fetchgits (because it got deleted  too early), so I  moved it to one level higher in the call stack, I think it's working now.

mihem commented on R/rix.R:
```diff
@@ -382,6 +383,10 @@ for more details."
     collapse = "\n"
   )
 
+  cache_dir <- file.path(tempdir(), "rix_cache")
```
yeah sorry, you are right, this line should not be necessary I think. But the following lines
```
if (dir.exists(cache_dir)) {
      unlink(cache_dir, recursive = TRUE)
    }
```
are necessary right now. I can see if i can remove this earlier if you like, but this did not work in a first attempt so I intentionally put it at the very end.

mihem commented on tests/testthat/_snaps/rix/mlr3_default.nix:
```diff
@@ -0,0 +1,160 @@
+      };
+    });
+
+
```
yeah sure should be doable. So make `ignore.cache` an argument of the main rix function?
Default should be `FALSE`, right? Should we change `skip_post_processing` to `TRUE` as default or leave it as `FALSE`?

b-rodrigues replied:
yes FALSE, for now let's leave skip_processing as is

b-rodrigues replied:
sorry make it ignore_cache with a _ not a .

mihem commented:
I can confirm, the devtools_test now correctly shows when it fails :D

mihem commented:
@b-rodrigues  I think i am more or less done.

But uff that was more difficult than I thought.

1) So removing is done now through `on.exit` in `fetchpkgs`. This works well when calling `rix`. You have to be careful when designing test and I had to adjust my tests. YOu have to initalze the cache file and delete it again.
```
cache_file <- get_cache_file()
...
  on.exit(unlink(cache_file))
```

I had some difficulties with placing the `on.exit` at the correct location, but now all tests pass.

2) Making `ignore_cache` work required quite some arguments to be added because it has to be passed throught so many functions. I tested it a debugger and also added a `test-dup.yaml`, which has `skip_post_processing = TRUE` and `ignore_cache = TRUE` and the resulting `default.nix` has  two `set6` packages. And on the other hand there's still the `test-dedup` which succeeds in removing the duplicates when using the cache (without using post-processing).
BTW: i moved the default.nix for these github workflow, not used for devtools testing, from `_snap` to `testdata` because after running `devtools::test`, there were automatically deleted (because not used by devtools).

Also keep in mind to change the link in `test-dedup.yaml` and test-dup.yaml`.


I was not sure about these lines:

```
     "check_nix" = nix_hash(repo_url, commit, ignore_cache),
      "locally" = nix_hash(repo_url, commit, ignore_cache),
      "api_server" = nix_hash_online(repo_url, commit)
```

Is that correct? Did not really understand how nix_hash_online work and

mihem commented:
Oh no.

I don't agree that this is a corner case. I think this is a quite common case because it's not so rare that Github package rely on other Github packages (I've experienced that and the other user as well).

Can you have a look again? I don't think it as thaaat complex, but I agree that there are quite some changes (also due to your idea to implement the `ignore_cache`, which I think is a good idea in general, but requires passing this argument through a lot of functions, so a lot of functions require change). 

And the problem remains that `post_processing` is suboptimal and requires more API calls than this solution and did not solve the problem with duplicates (only in combination with my code).

So maybe at second sight you see that this is not as complex as it seems, or see ways to simplify the code.

b-rodrigues commented:
i’ll have to think about it more. Perhaps we should use `...` in the functions instead of explicitely having `ignore_cache` that could already be something.

mihem commented:
So I thought about this again.

In the end, you should decide what's best for the package.

However, I think that this feature is quite important as soon as remote packages are involved because this is not such an uncommon situation. In contrast, to my previous two attempts and also post-processing, this solution solves all duplicate problems and avoids unnecessary GitHub API calls (which are limited and can also take time if many commits have to be downloaded). 
But the question remains whether this code could be simplified. I think getting the correct date via API calls is more complex and less robust (although I hope we made it robust with the tryCatch statements). By removing `ignore_cache` there would be much less code change, but I am not sure if this would really simplify the code, and I agree with you that it's convenient to deactivate this feature (for testing if this causes problems).
But maybe do you see other ways to simplify the code.
Or the other way around: Where do you see complexity in the code that you think might break that you would like to change?

b-rodrigues commented:
I did some changes to your PR: https://github.com/mihem/rix/pull/6 can you merge please?

mihem commented:
yes also just saw this ;)

mihem commented on R/fetchers.R:
```diff
@@ -610,43 +627,83 @@ get_closest_commit <- function(commits_df, target_date) {
+    }
+  }
+
+  cache_key <- if (length(remote_pkg_name_and_ref) == 2) {
```
yeah, you are right, makes more sense if it's in the if clause below

mihem commented:
Hm, only need to debug now why this stupid test_dup is failing now again.

mihem commented:
Hm, ignore_remotes_fache is not passed down to fetchgits. It's FALSE although it is specified as TRUE by the user.

mihem commented:
Hm I though the problem was that this
```
  args <- list(...)
  ignore_remotes_cache <- if (!is.null(args$ignore_remotes_cache)) args$ignore_remotes_cache else FALSE
```
is missing in `generate_git_archived_pkgs`

but actually the args are not named, so this code is not getting the value of ignore_remotes_cache (which should be TRUE in this case).

![image](https://github.com/user-attachments/assets/ba6a45b5-980a-4e45-97a5-df0e3bc04c4c)

mihem commented:
Hm okay, now there were are 2 problems.

1. Problems was introduced due to my last change. When using the fetchgit lookup test, the fetching was failing because the url provided in the description does not exist and this leads to `cache_key` not being assigned. Previously, `cache_key` was initalized as `NULL`, so did not fail then even if the fetching of the commits was not working. So i now changed this back and improved the docs. 

2. This is more complex. Due to the changes of using `...` instead of explictely naming `ignore_remotes_cache`, this argument is not passed on through the chain call. See my previous post. If the user provides `ignore_remotes_cache = TRUE` args is just `TRUE`, but `args$ignore_remotes_cache` is `NULL` so `ignore_remotesÄcache` is set to `FALSE` although it should be `TRUE`. Any ideas how to fix that?

mihem commented:
Oh man, I was playing around with `match.call` `substitute` `eval` and others, and the solution was easy as using

 `ignore_remotes_cache = ignore_remotes_cache`

in `generate_git_archived_pkgs`.

Now all test-dedup and test-dup pass :100:  
So hopefully ready for merging. @b-rodrigues

mihem commented:
@b-rodrigues anything you want me to do on this? rebase or something else?

b-rodrigues commented:
reviewing is on my todo

b-rodrigues commented on R/fetchers.R:
```diff
@@ -610,43 +627,88 @@ get_closest_commit <- function(commits_df, target_date) {
+
+  # Store package name and ref in cache key if ref is provided
+  # otherwise set to NULL
+  if (!ignore_remotes_cache) {
```
I don't understand this part, namely the == 2 and etc. What are you testing for? I think it would important to have comments here, and perhaps even illustrate with an example, if at all possible

mihem replied:
Ah that simply means that a ref is provided. 
Not really sure how to improve the comment "Store package name and ref in cache key if ref is provided otherwise set to NULL"?
I can provide an example, which is schex@031320d from
https://github.com/satijalab/seurat-wrappers/blob/master/DESCRIPTION
which is then converted in
https://github.com/ropensci/rix/blob/11fca1ce09e73cda45d97fe18be9050142dc3d48/R/fetchers.R#L238C4-L239C76
a list of two "scheck" and "031320d"

Does it help if I mention schex@031320d as an example in the comment?

b-rodrigues replied:
yeah definitely, if you could "walk the reader through" the whole process. such as "Description has remotes schex@ararar so the list will have two parts "schex" and the commit "ararar" so we check ..."

mihem replied:
okay will try to do this this evening

b-rodrigues commented on R/rix.R:
```diff
@@ -457,28 +469,25 @@ for more details."
 
 
 #' @noRd
-post_processing <- function(default.nix, flag_git_archive, skip_post_processing){
```
I think we need to decide what to do with post_processing? Do we keep it for removing duplicates or not? I think we should keep it at least for removing superfluous blank lines, but perhaps not for dedups anymore. Or should we keep it "just in case", as an added safety valve?

mihem replied:
I think these large blocks of superfluous blank lines actually resulted from post_processing so I think this is not strictly necessary.
I am also not sure, I would think/hope that post_processing is not necessary, but you spent some effort develeoping it and it's at least partly working, so maybe leave it as a safety net? Your decision.

mihem replied:
So it'f it's just about removing dobule empty lines we could also use your own liner, but actually I think it looks okay as it is. Not sure how it's recommended in Nix style guidelines.

mihem replied:
i think this is solved see below

b-rodrigues commented:
I've added some comments, I think that once these are addressed, you could `devtools::document()` and then if you could `devtools::check_win_devel()` that'd be great. If everything is ok, I think this is good to merge.

mihem commented on R/nix_hash.R:
```diff
@@ -262,19 +265,20 @@ nix_hash_online <- function(repo_url, commit) {
-      "locally" = nix_hash(repo_url, commit),
+      "check_nix" = nix_hash(repo_url, commit, ...),
+      "locally" = nix_hash(repo_url, commit, ...),
       "api_server" = nix_hash_online(repo_url, commit)
```
ah okay. But is this realted to this commit, which is avoiding dupliactes? Does nix_has_online also lead to downloading git hub packages and then using the imports from description?
If yes, I think I would need your help with setting up a integreation test.

b-rodrigues replied:
basically just generate the expression that would have duplicates in a runner on which you don't install nix. This should generate the valid expression nonetheless

mihem replied:
So this means just skipping 
```
      - name: Install Nix
        uses: DeterminateSystems/nix-installer-action@main
``` 
or also skipping 

```
  - uses: cachix/cachix-action@v15
        with:
          name: rstats-on-nix
          # If you chose signing key for write access
          # signingKey: '${{ secrets.CACHIX_SIGNING_KEY }}'
          # If you chose API tokens for write access OR if you have a private cache
          authToken: '${{ secrets.CACHIX_AUTH }}'
```

or anything else?

b-rodrigues replied:
well, the issue is that we would need to install R and rix without nix then. So yeah, you need to skip those, and then use other actions to install R. The other way would be to still use Nix, but then to set the option options(rix.sri_hash=x) where x is "api_server". But I think it would be better to skip installing nix entirely, to make extra sure this works even when nix is not present at all

mihem replied:
Sorry, don't understand the entire thing yet and not super experienced with running nix via github actions

Do you have a template for that?
If not, what do you think about?

```
steps:
- uses: actions/checkout@v4
- uses: r-lib/actions/setup-r@v2
  with:
    r-version: '3.5.3' # The R version to download (if necessary) and use.
    # Use "renv" to retrieve R version recorded in renv.lock file.
- run: Rscript -e 'install.packges("rix")'
```

Can I use cachix?

``` 
 - uses: cachix/cachix-action@v15
        with:
          name: rstats-on-nix
          # If you chose signing key for write access
          # signingKey: '${{ secrets.CACHIX_SIGNING_KEY }}'
          # If you chose API tokens for write access OR if you have a private cache
          authToken: '${{ secrets.CACHIX_AUTH }}'
```

and then 

```
Rscript -e \"devtools::load_all();rix(date = '2023-12-30', r_pkgs = c('tidyverse', 'mlr3', 'qs', 'matrixStats', 'prospectr', 'Cubist', 'checkmate', 'mlr3misc', 'paradox'), git_pkgs = list(package_name = 'mlr3extralearners', repo_url = 'https://github.com/mlr-org/mlr3extralearners/', commit = '6e2af9ef9ecd420d2be44e9aa2488772bb9f7080'), ide = 'none', project_path = '.', overwrite = TRUE, print = FALSE, skip_post_processing = TRUE, ignore_remotes_cache = FALSE)"
```
still works although nix is not installed?

b-rodrigues replied:
no cachix needed, but the first action to install R is required 

the idea is that rix generates expressions even if nix is not present on the system, but the issue is with the hash. So we set up a server that computes the hash and sends it back. We just need to make sure this mechanism is still working in the context of this PR

mihem replied:
i think this is more complex, not directly linked to this PR (removing duplicated), so I would prefer if this would be a different PR

mihem commented:
@b-rodrigues i improved comments and removed post_processing. I think removing `post_processing` resulted in all test passing. I only had to adjust several snapshots  by adding one blank line, e.g.:

![image](https://github.com/user-attachments/assets/d47437e6-7721-4259-a0fd-88e3a07375ce)


I think this looks good. Agree?

Still working on the nix_hash_online issue.

mihem commented:
Hm I created a test-dedup-nix-hash-online.yaml workflow. I think this is setup correctly?

I failed with 

```
ℹ Loading rix
Nix doesn't seem to be installed on this system.
You can still generate Nix expressions, but you won't be able to build them.
Error in order(sapply(git_pkgs, "[[", "package_name")) : 
  unimplemented type 'list' in 'orderVector1'
Calls: rix ... fetchgit -> fetchgits -> [ -> [.data.frame -> order
Execution halted
```

So me and Copilot thought this was a problem of sapply, so better use vapply? 
However this did not fix things. Now this error occur:

```
ℹ Loading rix
Nix doesn't seem to be installed on this system.
You can still generate Nix expressions, but you won't be able to build them.
Error in vapply(git_pkgs, function(x) x$package_name, character(1)) : 
  values must be length 1,
 but FUN(X[[1]]) result is length 0
Calls: rix ... paste -> fetchgits -> fetchgit -> fetchgits -> vapply
Execution halted
Error: Process completed with exit code 1.
```

mihem commented:
Ah the problem is that the output of `output$deps$remotes` looks like this:

`options(rix.sri_hash = "api_server")`

![image](https://github.com/user-attachments/assets/a9975102-c0fa-4df5-9587-79b6c88e1ad6)


while when options is not specified on my system, so uses `nix_hash)` then it' looks like this:

![image](https://github.com/user-attachments/assets/d9b60c55-b7cb-4466-bdcd-f597abaf3ec5)

So first of all the structure is wrong and probably causes this problem and second of all it just fallsback to `HEAD` (is this okay?
I am pretty sure that this has nothing to do with this PR (removing duplicates by using caching), so I would prefer if we would merge this PR first and then work on that one in a a new PR. This one is already big enough I think ;).

mihem commented:
I've run  `devtools::check_win_devel()` but also got na error that pdflatex was missing. So installed tinytex (and tinytex within tinytex :P), but then got the following error:

`
Upload failed (at start/before it took off) [win-builder.r-project.org]: Failed FTP upload: 550
`

b-rodrigues commented:
I'm not sure yet if it's not linked but in any case I'll have to fix it first then

b-rodrigues commented:
> > I am pretty sure that this has nothing to do with this PR (removing duplicates by using caching), so I would prefer if we would merge this PR first and then work on that one in a a new PR. This one is already big enough I think ;).
> 
> I'm not sure yet if it's not linked but in any case I'll have to fix it first then

ok so I’ve added an action to run R CMD Check on a runner without Nix, and indeed, there is something wrong the response from the server. I thiiiink I know what’s going on:

https://github.com/ropensci/rix/actions/runs/13562423321/job/37908068713

b-rodrigues commented:
I realized `  testthat::expect_snapshot_file` is skipped on CRAN by default, so these tests are not running through the rhub action either. So yeah, this might have been broken for a long time actually

b-rodrigues commented:
I want to deal with this issue first: https://github.com/ropensci/rix/issues/430

I think that the bug was introduced with the changes from the last PR

b-rodrigues commented:
Could you resolve conflicts?

mihem commented:
Will try to do this during work on GitHub directly, if it does not work I will do it this evening.

mihem commented:
oh man this really sucks online. I resolved all conflcit on github, but then "commiting" does not work, I then tried to copy manually each file and commiting manually, which is also not great. have to do this this evening in VSCode.

b-rodrigues commented:
ready to merge ?

mihem commented:
give me 5 more mintues to manually check.

It was again quite confusing because i didn't realize that the resolving conflicts removed the ..., so ignore_remotes_cache was not passed, and this produced some weird error.

mihem commented:
Yeah I think everything is good! Long journey! :+1:

mihem commented:
or let me write this caching a little in the docs

b-rodrigues commented:
yes that'd be great I'll wait then

mihem commented:
Have a look if you are okay with the explaination.
Found it difficult to explain the issue, especially for the "correct" dates.

Otherwise ready to merge finally.

b-rodrigues commented on vignettes/z-advanced-topic-handling-packages-with-remote-dependencies.Rmd:
```diff
@@ -206,6 +206,26 @@ in `Remotes` in the DESCRIPTION file will work.
+However, if you disable this feature, then `{rix}` will fetch package `C` twice, once for the commit of the 1st December 2024 as before
+(closest date that is earlier than the commit date of package `A`) and once for a date
+before the 1st November 2024 (closest date that is earlier than the commit date of package `B`).
+You should be aware of this because if you are unlucky the default option `ignore_remotes_cache = FALSE` might result in a solution that does not work.
```
you mean it might not work because the commit that will be picked might be for a version of C incompatible with either A or B?

mihem replied:
Yes. We observed this with he mlr3 example.
I mean this can always happen with GitHub packages. I just wanted the user to be aware of the fact that there a multiple solution of which commit rix should pick for package c in sich a situation and that some of them might work and others not. And that the one rix chooses is the one that is closest to package a, but not necessarily the one that works.
Maybe you find better words.

b-rodrigues commented:
Many thanks! this is a great addition. I was a bit worried at first since it introduced quite a lot of code, but it's for the better.

mihem commented:
Welcome.
Also think this is an important addition for handling remote dependencies. It hopefully solves the duplicate problem for good. We could remove my previous two loops to remove duplicates and the post_processing function.
I hope that now, with fetching the closest commit and caching packaces, fetching remote dependencies from GitHub is more or less stable.

I saw that you edited the links for the tests workflows. However, you also need to change the branch, not only the username.
